proposals/windows-mdm-commands-tables.md

```diff
@@ -0,0 +1,118 @@
+# Changes to Windows MDM command tables
+
+**Context** Commands are the mechanism we'll use to build features like
+
+- Deliver OS settings
+- Filter hosts by their "OS settings" status (pending, verified, verifying, failed)
+
+IF possible I would like us to have the same mental model for both platforms,
+where we have three tables with a clear separation of concerns:
+
+A. List of commands
+B. Queue
+C. List of command results
+
+Any command, doesn't matter how it was enqueued (via the UI/CLI or in response
+to other commands) should have an entry in `A`, plus entries with FK references
+in `B` when it's enqueued, and `C` when we get results.
+
+The [current schema](https://github.com/fleetdm/fleet/blob/defad14dd92ac1fb049552ec1564bceb50606e2d/server/datastore/mysql/migrations/tables/20230926082152_AddWindowsMDMTables.go)
+for Windows has two tables, with the following descriptions:
+
+### `windows_mdm_pending_commands`
+
+> The table windows_mdm_pending_commands contains the stateless commands enqueued into the MDM pending commands table from fleetctl or fleet Server.
+> These commands contain operations to be performed on a given device. 
+> They don't yet have the MessageID, SessionID, and CmdId information as this data gets populated during the response 
+> to a given MDM management request.
+
+### `windows_mdm_commands`
+
+> The windows_mdm_commands table represents the stateful commands that were actually sent over the wire. 
+> These commands include the MessageID, SessionID, and CmdId as they were sent to the device. 
+> They are stored in the commands table for error code tracking purposes. 
+> Additionally, this table can contain commands that were not part of the windows_mdm_pending_commands table and 
+> were generated by the Fleet server in response to an incoming MDM command.
+
+## Proposal
+
+I think it's fundamental we have:
+
+- A clear queue model
+- A single source of truth for commands
+- The status of a command at any point in time easily available
+- The raw XML we sent to the device
```

Comment:
I agree that it would be simpler to store the raw XML instead of deconstructing the relevant fields (and presumably reconstructing when we send it back to the fields, I haven't seen that part yet). It might've been proposed like that due to not having a full raw XML for server-initiated commands, but nothing would prevent us from generating it and saving it instead of generating only when we need to send the commands.

Reply:
Yes, we don't have the full raw XML for server-initiated commands. A SyncML command sent from the server to the host contains not only the protocol commands (Replace, Get, etc), but also the Status/Results responses for commands sent from the device to the server. Also, the final values for MessageID, CmdID, and SessionID fields are only known at server response time.

Reply:
I also wonder if the commands that the device can send are quite limited and could be special cased? In Apple land, "device commands" are special protocol messages (eg: `TokenUpdate`) and we treat those differently

Reply:
> I also wonder if the commands that the device can send are quite limited

Yes, these type of commands are limited, and we don't have to track them on the DB, as protocol command responses are meant to comply with the protocol.  The server generates status/Results protocol command responses, which are ACK responses to the MDM client stack.

Reply:
One important clarification for future reference:
- `SyncML message`: The MS-MDM message sent over the wire. It is composed of a header + body. SyncML commands are contained in the SyncML body. MessageID/SessionID fields are set here.
-  `SyncML command`: The actual operation to be performed on the device. This is also called `protocol command`.  An operation verb defines the action to perform on the device(Get, Replace, etc), and the rest of the fields in the protocol command payload establish the setting (OMA-URI) to modify. `CommandID` field is set here.

@mna @roperzh - I've documented SyncML message request/response examples with different types of SyncML commands [here](https://docs.google.com/document/d/19OClzd3yaVqlrgVSw2oy6152NykYLLEEf3u-i_EZhek/edit?usp=sharing). 

One of the tests I made last week was checking if CommandID of a protocol command can be set to a UUID, and I found this was possible (see section `Checking if it possible to use arbitrary content in Command ID`)

This means we can potentially avoid requiring the MessageID/SessionID fields from the container and rely on the CommandID to track protocol command responses. This would be inline with this proposal with some minor modifications

Reply:
> One of the tests I made last week was checking if CommandID of a protocol command can be set to a UUID, and I found this was possible 

@marcosd4h  That sounds amazing and a great find! Would simplify our logic a bit. 

> I think this is a key aspect, devices can also send commands to the server.

@roperzh Yeah this is a major distinction with Apple's protocol.

> I think that means that the current proposal needs adjustments, but I'm not sure what we need to do yet

+1, same here. Thinking out loud, looks like we could still store the raw xml of the commands in the tables, and the server could unmarshal those to the structs that Marcos created and add any required response to the device dynamically. The tables would track server-initiated commands only (including user-sent commands through `fleetctl mdm run-command`), while the device-initiated commands (which AIUI are really similar to `TokenUpdate` and related commands required for the "maintenance" of the MDM protocol between device and server) would not get stored (just as we don't store those calls in Apple-world)?

> Also, the final values for MessageID, CmdID, and SessionID fields are only known at server response time.

Those are really protocol-internal things though, so it looks like again those could be filled dynamically at server response time and not a big deal if they are not part of the raw xml saved in the tables? And when we know them, we could store them as fields in the DB so the actual XML sent could be regenerated if needed.

Question: in case of failure sending the response, on retry do those message/cmd/session IDs stay the same or get new values (presumably session ID gets a new value)?

Reply:
>Question: in case of failure sending the response, on retry do those message/cmd/session IDs stay the same or get new values (presumably session ID gets a new value)?

yes, in the case of a retry, the message/session IDs will be regenerated as they belong to the SyncML message  and this is something that changes over time. On the other hand, the CmdID of the actual protocol operation can be the same

Reply:
In that case maybe we don't even need to store them in the DB and just fill them dynamically every time, and we'll have them as part of the result anyway (in the raw XML). (maybe that's what you meant here, apologies if I'm just now catching up :D :

> we can potentially avoid requiring the MessageID/SessionID fields from the container and rely on the CommandID to track protocol command responses.